### ISSUE TITLE
Prevent Git concurrency

### ISSUE THREAD

**Issue by MattiSG** [ambanum/OpenTermsArchive#753]
There is an increasing amount of cases where a lock file cannot be obtained on the Git repository and where the process crashes. Meanwhile, versions are still created and pushed.

Below are example error logs and a count of each error type.

Observations:

- It is unexpected that multiple Git or OTA instances are run in parallel. It has been demonstrated that only a single OTA instance exists on the faulty servers (stopping it stops all change trackings, and `ps` shows that there is no other Node process running).
- Since the failures appear on random services, it is impossible that the lock files are leftovers that are not cleaned.

### 28/02/2022

Many errors during the week end
```
2022-02-28 06:30:15 �[31merror�[39m Amazon — Privacy Policy Error: Could not commit /home/debian/ota-versions-france/Amazon/Privacy Policy.md with message "Update Amazon Privacy Policy

This version was recorded after filtering snapshot https://github.com/OpenTermsArchive/snapshots-france/commit/621c6bf6b5b4f9ba1cab1611" due to error: "Error: fatal: Unable to create '/home/debian/ota-versions-france/.git/index.lock': File exists.

Another git process seems to be running in this repository, e.g.
an editor opened by 'git commit'. Please make sure all processes
are terminated then try again. If it still fails, a git process
may have crashed in this repository earlier:
remove the file manually to continue.
"
at GitAdapter._commit (file:///home/debian/ota/src/storage-adapters/git/index.js:99:13)
at async GitAdapter.record (file:///home/debian/ota/src/storage-adapters/git/index.js:35:17)
at async Archivist.recordVersion (file:///home/debian/ota/src/archivist/index.js:231:46)
at async Archivist.trackDocumentChanges (file:///home/debian/ota/src/archivist/index.js:147:29)
```
#### actions taken

@martinratinaud: restart services
- ansible-playbook --inventory ops/inventories/production.yml ops/app.yml --limit france  --tag restart

### 24/02/2022

Seems to be fine now, only one error email

**3 errors**
```
2022-02-24 06:30:09 �[31merror�[39m Amazon — Terms of Service Error: Could not commit /home/debian/ota-versions-france/Amazon/Terms of Service.md with message "Update Amazon Terms of Service

This version was recorded after filtering snapshot https://github.com/OpenTermsArchive/snapshots-france/commit/621725f005318ee21546c454" due to error: "Error: fatal: Unable to create '/home/debian/ota-versions-france/.git/index.lock': File exists.

Another git process seems to be running in this repository, e.g.
an editor opened by 'git commit'. Please make sure all processes
are terminated then try again. If it still fails, a git process
may have crashed in this repository earlier:
remove the file manually to continue.
"
at GitAdapter._commit (file:///home/debian/ota/src/storage-adapters/git/index.js:99:13)
at async GitAdapter.record (file:///home/debian/ota/src/storage-adapters/git/index.js:35:17)
at async Archivist.recordVersion (file:///home/debian/ota/src/archivist/index.js:231:46)
at async Archivist.trackDocumentChanges (file:///home/debian/ota/src/archivist/index.js:147:29)
```

```
2022-02-24 00:31:55 �[31merror�[39m La Redoute — Privacy Policy Error: Could not commit /home/debian/ota-versions-france/La Redoute/Privacy Policy.md with message "Update La Redoute Privacy Policy

This version was recorded after filtering snapshot https://github.com/OpenTermsArchive/snapshots-france/commit/6216d1fa7e09a80c90655fbd" due to error: "Error: fatal: Unable to create '/home/debian/ota-versions-france/.git/index.lock': File exists.

Another git process seems to be running in this repository, e.g.
an editor opened by 'git commit'. Please make sure all processes
are terminated then try again. If it still fails, a git process
may have crashed in this repository earlier:
remove the file manually to continue.
"
at GitAdapter._commit (file:///home/debian/ota/src/storage-adapters/git/index.js:99:13)
at processTicksAndRejections (internal/process/task_queues.js:95:5)
at async GitAdapter.record (file:///home/debian/ota/src/storage-adapters/git/index.js:35:17)
at async Archivist.recordVersion (file:///home/debian/ota/src/archivist/index.js:231:46)
at async Archivist.trackDocumentChanges (file:///home/debian/ota/src/archivist/index.js:147:29)
```

```
2022-02-24 00:30:00 �[31merror�[39m unhandledRejection: error: could not lock config file /home/debian/ota-versions-france/.git/config: File exists
fatal: could not set 'core.filemode' to 'true'

Error: error: could not lock config file /home/debian/ota-versions-france/.git/config: File exists
fatal: could not set 'core.filemode' to 'true'

at Object.action (/home/debian/ota/node_modules/simple-git/src/lib/plugins/error-detection.plugin.js:30:33)
at PluginStore.exec (/home/debian/ota/node_modules/simple-git/src/lib/plugins/plugin-store.js:21:33)
at /home/debian/ota/node_modules/simple-git/src/lib/runners/git-executor-chain.js:95:45
at new Promise ()
at GitExecutorChain.handleTaskData (/home/debian/ota/node_modules/simple-git/src/lib/runners/git-executor-chain.js:93:16)
at GitExecutorChain. (/home/debian/ota/node_modules/simple-git/src/lib/runners/git-executor-chain.js:77:46)
at Generator.next ()
at fulfilled (/home/debian/ota/node_modules/simple-git/src/lib/runners/git-executor-chain.js:5:58)
```
### 18/02/2022 -> 21/02/2022
~Incident ongoing 18/02/2022 → 21/02/2022, total of 11 errors.~
Incident fixed

#### actions taken

@martinratinaud: restart services
- ansible-playbook --inventory ops/inventories/production.yml ops/app.yml --limit services_all  --tag restart
- ansible-playbook --inventory ops/inventories/production.yml ops/app.yml --limit france  --tag restart

#### Lockfile present

6 occurrences

Example:

```
2022-02-21 00:31:28 [31merror[39m Snapchat — Terms of Service Error: Could not commit /home/debian/ota-versions-france/Snapchat/Terms of Service.md with message "Update Snapchat Terms of Service

This version was recorded after filtering snapshot https://github.com/OpenTermsArchive/snapshots-france/commit/6212dd5fcb03a20e6d40c01f" due to error: "Error: fatal: Unable to create '/home/debian/ota-versions-france/.git/index.lock': File exists.
```

Note: it happened once that the commit reference is `true` instead of a snapshot ID.

```
2022-02-20 18:30:28 [31merror[39m Canalplus — Terms of Service Error: Could not commit /home/debian/ota-versions-france/Canalplus/Terms of Service.md with message "Update Canalplus Terms of Service

This version was recorded after filtering snapshot https://github.com/OpenTermsArchive/snapshots-france/commit/true" due to error: "Error: fatal: Unable to create '/home/debian/ota-versions-france/.git/index.lock': File exists.
```

#### Config file lock

3 occurrences

Example:

```
2022-02-20 18:30:00 [31merror[39m unhandledRejection: error: could not lock config file /home/debian/ota-versions-france/.git/config: File exists
fatal: could not set 'core.repositoryformatversion' to '0'

Error: error: could not lock config file /home/debian/ota-versions-france/.git/config: File exists
fatal: could not set 'core.repositoryformatversion' to '0'
```

```
2022-02-21 12:30:00 [31merror[39m unhandledRejection: error: could not lock config file /home/debian/ota-versions-france/.git/config: File exists
fatal: could not set 'core.bare' to 'false'

Error: error: could not lock config file /home/debian/ota-versions-france/.git/config: File exists
fatal: could not set 'core.bare' to 'false'
```

#### Remote not up to date

2 occurrences

Example:

```
2022-02-20 06:31:43 [31merror[39m unhandledRejection: To github.com:OpenTermsArchive/versions-france.git
! refs/heads/main:refs/heads/main [remote rejected] (cannot lock ref 'refs/heads/main': is at 63614296123266570674f4bc7b7479de23e6ee9a but expected b963545bec2e91f49bb5c72f1c03a9444172975e)
Done
Pushing to github.com:OpenTermsArchive/versions-france.git
error: failed to push some refs to 'github.com:OpenTermsArchive/versions-france.git'

Error: To github.com:OpenTermsArchive/versions-france.git
! refs/heads/main:refs/heads/main [remote rejected] (cannot lock ref 'refs/heads/main': is at 63614296123266570674f4bc7b7479de23e6ee9a but expected b963545bec2e91f49bb5c72f1c03a9444172975e)
Done
Pushing to github.com:OpenTermsArchive/versions-france.git
error: failed to push some refs to 'github.com:OpenTermsArchive/versions-france.git'
```

**Comment by MattiSG** [comment]
Isn't this fixed by https://github.com/ambanum/OpenTermsArchive/pull/766? What are the conditions for closing this?

**Comment by Ndpnt** [comment]
Yes, I think it should be fixed by #766. If the problem occurs again, we will open a new issue.

**Comment by MattiSG** [comment]
Is it this fix that closed automatically so many issues in France yesterday? 🙂

**Comment by Ndpnt** [comment]
I'm not sure.

**Comment by martinratinaud** [comment]
Do not know either and did not find so many closed yesterday (Only 2 https://github.com/OpenTermsArchive/declarations-france/issues?q=is%3Aissue+is%3Aclosed) or am I mistaking?

**Comment by martinratinaud** [comment]
Well, looks good then!
